File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
 class RandomizedSet {
    private int[] values;
    private int[] indexMap;
    private boolean[] present;
    private int size;
    private final int CAPACITY = 100003; // large prime to avoid collisions
    private Random rand;

    public RandomizedSet() {
      values = new int[CAPACITY];
      indexMap = new int[CAPACITY];
      present = new boolean[CAPACITY];
      size = 0;
      rand = new Random();
    }

    public boolean insert(int val) {
      int hash = hash(val);
      if (present[hash])
        return false;
      values[size] = val;
      indexMap[hash] = size;
      present[hash] = true;
      size++;
      return true;
    }

    public boolean remove(int val) {
      int hash = hash(val);
      if (!present[hash])
        return false;
      int idx = indexMap[hash];
      int lastVal = values[size - 1];
      values[idx] = lastVal;
      indexMap[hash(lastVal)] = idx;
      present[hash] = false;
      size--;
      return true;
    }

    public int getRandom() {
      return values[rand.nextInt(size)];
    }

    private int hash(int val) {
      return (val % CAPACITY + CAPACITY) % CAPACITY;
    }
  }

  class RandomizedSet1 {
    private int[] values;
    private int[] indexMap;
    private boolean[] present;
    private int size;
    private final int CAPACITY = 100003; // large prime to avoid collisions
    private Random rand;

    public RandomizedSet1() {
      values = new int[CAPACITY];
      indexMap = new int[CAPACITY];
      present = new boolean[CAPACITY];
      size = 0;
      rand = new Random();
    }

    public boolean insert(int val) {
      int hash = hash(val);
      if (present[hash])
        return false;
      values[size] = val;
      indexMap[hash] = size;
      present[hash] = true;
      size++;
      return true;
    }

    public boolean remove(int val) {
      int hash = hash(val);
      if (!present[hash])
        return false;
      int idx = indexMap[hash];
      int lastVal = values[size - 1];
      values[idx] = lastVal;
      indexMap[hash(lastVal)] = idx;
      present[hash] = false;
      size--;
      return true;
    }

    public int getRandom() {
      return values[rand.nextInt(size)];
    }

    private int hash(int val) {
      return (val % CAPACITY + CAPACITY) % CAPACITY;
    }
  }

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 1-1000/0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
// class RandomizedSet {
// public:
//     vector<int> nums;
//     unordered_map<int,int> mp;

//     RandomizedSet() {}

//     bool insert(int val) {
//         if(mp.count(val)) return false;
//         nums.push_back(val);
//         mp[val] = nums.size() - 1;
//         return true;
//     }

//     bool remove(int val) {
//         if(!mp.count(val)) return false;

//         int idx = mp[val];            // index of val
//         int last = nums.back();       // last element

//         nums[idx] = last;             // swap with last
//         mp[last] = idx;               // update index of last

//         nums.pop_back();              // remove last
//         mp.erase(val);                // remove val from map

//         return true;
//     }

//     int getRandom() {
//         int idx = rand() % nums.size();
//         return nums[idx];
//     }
// };


class RandomizedSet {
public:

vector<int>list;
unordered_map<int,int>map;

    RandomizedSet() {
         
    }
    
    bool insert(int val) {
        if(map.count(val)){
            return false;
        }
            list.push_back(val);
          map[val]=list.size()-1;
            return true;
    
    }
    
    bool remove(int val) {
        if(!map.count(val)) {
            return false;
        }

        int idx=map[val];
        int last=list.back();

        //swap with last
        list[idx]=last;
        map[last]=idx;

        
        list.pop_back();
        map.erase(val);
        return true;
    }
    
    int getRandom() {
      return list[rand()% list.size()];
    }
};
auto init = atexit([]{ofstream("display_runtime.txt")<<"0";});
/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */